src/db.rs

```diff
@@ -1287,6 +1449,241 @@ mod tests {
+        let object_store: Arc<dyn ObjectStore> = Arc::new(InMemory::new());
+        let db = Db::open_with_opts(
+            Path::from("/tmp/test_kv_store"),
+            test_db_options(0, 1024, None),
+            object_store,
+        )
+        .await
+        .unwrap();
+
+        seed_database(&db, &table, false).await;
+
+        if await_durable {
+            db.flush().await.unwrap();
+        }
+
+        db
+    }
+
+    async fn assert_empty_scan(db: &Db, range: BytesRange) {
+        let mut iter = db
+            .inner
+            .scan_with_options(range.clone(), &ScanOptions::default())
+            .await
+            .unwrap();
+        assert_eq!(None, iter.next().await.unwrap());
+    }
+
+    #[test]
+    fn test_empty_scan_range_returns_empty_iterator() {
+        let mut runner = new_proptest_runner(None);
+        let table = sample::table(runner.rng(), 1000, 5);
+
+        let runtime = Runtime::new().unwrap();
+        let db = runtime.block_on(build_database_from_table(&table, true));
+
+        runner
+            .run(&arbitrary::empty_range(10), |range| {
+                runtime.block_on(assert_empty_scan(&db, range));
+                Ok(())
+            })
+            .unwrap();
+    }
+
+    async fn assert_records_in_range(
+        table: &BTreeMap<Bytes, Bytes>,
+        db: &Db,
+        scan_options: &ScanOptions,
+        range: BytesRange,
+    ) {
+        let mut iter = db
+            .inner
+            .scan_with_options(range.clone(), scan_options)
+            .await
+            .unwrap();
+        assert_ordered_scan_in_range(&table, &range, &mut iter).await;
+    }
+
+    #[test]
+    fn test_scan_returns_records_in_range() {
+        let mut runner = new_proptest_runner(None);
+        let table = sample::table(runner.rng(), 1000, 5);
+
+        let runtime = Runtime::new().unwrap();
+        let db = runtime.block_on(build_database_from_table(&table, true));
+
+        runner
+            .run(&arbitrary::nonempty_range(10), |range| {
+                runtime.block_on(assert_records_in_range(
+                    &table,
+                    &db,
+                    &ScanOptions::default(),
+                    range,
+                ));
+                Ok(())
+            })
+            .unwrap();
+    }
+
+    fn new_proptest_runner(rng_seed: Option<[u8; 32]>) -> TestRunner {
+        let rng = rng::new_test_rng(rng_seed);
+        let mut config = proptest::test_runner::contextualize_config(Config::default().clone());
+        config.source_file = Some(file!().into());
+        TestRunner::new_with_rng(config, rng)
+    }
+
+    #[test]
+    fn test_scan_returns_uncommitted_records_if_read_level_uncommitted() {
+        let mut runner = new_proptest_runner(None);
+        let table = sample::table(runner.rng(), 1000, 5);
+
+        let runtime = Runtime::new().unwrap();
+        let db = runtime.block_on(build_database_from_table(&table, false));
```

Comment:
it would probably be good to set a long wal flush interval on this one to ensure we're reading nondurable rows